### ISSUE TITLE
Add terraform resource for Grafana Stacks

### ISSUE THREAD

**Issue by NasAmin** [grafana/terraform-provider-grafana#342]
### Problem
Grafana Cloud provides a [Stacks API](https://grafana.com/docs/grafana-cloud/reference/cloud-api/#stacks).
Customers using Grafana Cloud and want to manage multiple stacks/instances of Grafana currently rely on a manual process using the Grafana admin console. This is not sustainable for a scaling organisation.

Providing a terraform provider for creating new Grafana stacks will help users automate stack creation and closing the loop on manual processes.

### Proposal
Since there is already an API available, the following is proposed:

* Add support for the [Stacks API](https://grafana.com/docs/grafana-cloud/reference/cloud-api/#stacks) to the [Grafana API client for golang](https://github.com/grafana/grafana-api-golang-client)
* Use the above client library in this terraform provider and add a `grafana_cloud_stack` resource as well as a `grafana_cloud_stack` data source

ℹ️   I am happy to contribute

**Comment by inkel** [comment]
@NasAmin that's a great idea! Feel free to submit a PR for it 😺

**Comment by NasAmin** [comment]
@inkel Thanks. What would be the best way for testing any terraform changes locally for the [Stacks API](https://grafana.com/docs/grafana-cloud/reference/cloud-api/#stacks)? The README is not very clear on how to test the CloudAPI changes.

I am guessing I'll have to run acceptance tests against a real account on grafana.com?

**Comment by julienduchesne** [comment]
> I am guessing I'll have to run acceptance tests against a real account on grafana.com?

Yes, that's how I would do it. I can add a `GRAFANA_CLOUD_TOKEN` into our CI so you could run tests there if you like. We already have a test accounts we run tests in

**Comment by NasAmin** [comment]
@julienduchesne Thanks. 
That won't work on the fork, once I open the PR on the upstream repo, it will then use the `GRAFANA_CLOUD_TOKEN` correct?

**Comment by julienduchesne** [comment]
> @julienduchesne Thanks. That won't work on the fork right, once I open the PR on the upstream repo, it will then use the `GRAFANA_CLOUD_TOKEN` correct?

Yes. You're free to work in a draft PR to test your changes though!

**Comment by NasAmin** [comment]
@julienduchesne @inkel I have opened a [PR](https://github.com/grafana/grafana-api-golang-client/pull/53) on the grafana-api-golang-client repo. Please review 🙇

**Comment by julienduchesne** [comment]
> @julienduchesne @inkel I have opened a [PR](https://github.com/grafana/grafana-api-golang-client/pull/53) on the grafana-api-golang-client repo. Please review 🙇

Hey @NasAmin. It LGTM overall. I'd like to approve only when you have implemented the Terraform resource. The test framework in the go-client is insufficient, so I rely a lot on the Terraform provider tests which covers the client e2e. You can use a go.mod `replace` in your Terraform provider pull request to use your new go-client code

Also, I've just been informed that there's already a provider that supports stacks: https://github.com/form3tech-oss/terraform-provider-grafanacloud/blob/master/grafanacloud/resource_stack.go if you want to borrow some code

**Comment by NasAmin** [comment]
@julienduchesne Thanks, yes the test framework is insufficient. I have been testing the changes by making real api calls locally.

I have a working terraform implementation but thanks for linking the existing stacks provider. I'll take inspiration from it. I'll open a PR when ready, for the TF implementation.

**Comment by NasAmin** [comment]
@julienduchesne I have opened a [draft PR](https://github.com/grafana/terraform-provider-grafana/pull/354). However, some of my tests are [failing in CI](https://drone.grafana.net/grafana/terraform-provider-grafana/464/3/2). I am struggling to figure out why it is returning a `404 not found` error in the apply step.

I'll appreciate some guidance 🙇

**Comment by julienduchesne** [comment]
> @julienduchesne I have opened a [draft PR](https://github.com/grafana/terraform-provider-grafana/pull/354). However, some of my tests are [failing in CI](https://drone.grafana.net/grafana/terraform-provider-grafana/464/3/2). I am struggling to figure out why it is return a `404 not found` error in the apply step.
> 
> I'll appreciate some guidance 🙇

The issue is that the `gapi` client being created in `provider.go` is a Grafana client, not Grafana.com (cloud). The 404 you're hitting is because it is trying to query the /api/instances/ endpoint on a Grafana instance. 

To interact with the Cloud API, you'll need to plugin the `GRAFANA_CLOUD_API_KEY` env variable in `provider.go` to a `cloud_api_key` setting and create a new client. Something like this: https://github.com/grafana/terraform-provider-grafana/blob/master/grafana/provider.go#L93-L99

**Comment by NasAmin** [comment]
Thanks. So does this mean `gapi` is not the right client to use?
Shall I create a new client within the terraform provider or create the client in a new package?

**Comment by julienduchesne** [comment]
> Thanks. So does this mean `gapi` is not the right client to use? Shall I create a new client within the terraform provider or create the client in a new package?

Yeah, maybe the `grafana-api-golang-client` repo is not the right place for this. I think you could create a new client in the provider repo (read: the repo where we currently are) and forget about the `grafana-api-golang-client` repo. We (maintainers) can decide where we put it afterwards. The important part is the Terraform interface IMO

Sorry about the back and forth. Hadn't given any thought about that yet

**Comment by NasAmin** [comment]
@julienduchesne Just so I understand the issue:

When I test the provider locally, it works as expected i.e. 

```
export GRAFANA_URL=https://grafana.com
export GRAFANA_AUTH=<admin token>

provider "grafana" {
  url = "https://grafana.com"

  # export TF_VAR_grafana_auth=...
  auth = var.grafana_auth
}

resource "grafana_cloud_stack" "test" {
  name   = "myteststack"
  slug   = "myteststack"
  region_slug = "eu"
  description = "here is a test description"
}

data "grafana_cloud_stack" "test" {
  slug = grafana_cloud_stack.test.slug
}
```

The above works as expected in our cloud account.

However, the way the tests are run, the GRAFANA_URL is set to `https://terraformprovidergrafana.grafana.net/` which leads me to think the `/instances` api is not available on it.

Questions:
* Can we set `GRAFANA_URL` to grafana.com?
* Is grafana.com supported in CI currently?
* Is the `GRAFANA_CLOUD_API_KEY` for grafana.com and does it have permissions to create stacks?

I apologise but seems like we are getting into a design discussion here.

**Comment by NasAmin** [comment]
I have added a new pipeline for cloud stacks with the right url and token. However, the CI build is blocked. I understand this is because I have updated the `drone.yaml` file which needs approval?
![image](https://user-images.githubusercontent.com/12657747/151152706-96b59d11-3547-4779-ab39-71baef1a21ed.png)

**Comment by NasAmin** [comment]
@julienduchesne @inkel Hi guys

First of all, thank you so much for your quick and helpful responses.
I think the testing framework is currently a bit limited as the provider has a mix of cloud, OSS and enterprise resources. On top of that some resources are instance specific whilst others are for grafana.com.

Instead of going into a design discussion around splitting the provider into multiple providers, based on the type of resources, here's what I am requesting:
* Are the changes made to `drone.yaml` acceptable? I am unable to generate the `hmac` so I'll need your help 
**OR**
* Will you accept the terraform resources + data source code for `Stacks` without integration tests? 
**OR**
* Can we have a `skip` condition on the cloud stack integration tests on the [current PR](https://github.com/grafana/terraform-provider-grafana/pull/354) and you as maintainers can workout the best way to run those tests?
 
Thanks a lot!

**Comment by julienduchesne** [comment]
> @julienduchesne @inkel Hi guys
> 
> First of all, thank you so much for your quick and helpful responses. I think the testing framework is currently a bit limited as the provider has a mix of cloud, OSS and enterprise resources. On top of that some resources are for a specific instances whilst others are for grafana.com.
> 
> Instead of going into a design discussion around splitting the provider into multiple providers, based on the type of resources, here's what I am requesting:
> 
> * Are the changes made to drone.yaml acceptable? I am unable to generate the `hmac` so I'll need your help
>   **OR**
> * Will you accept the terraform resources + data source code for `Stacks` without integration tests?
>   **OR**
> * Can we have a `skip` condition on the cloud stack integration tests on the [current PR](https://github.com/grafana/terraform-provider-grafana/pull/354) and you as maintainers can workout the best way to run those tests?
> 
> Thanks a lot!

Thanks for your work @NasAmin. I'll discuss this with @inkel and try to make things easier for you. You are telling me that the current client implementation works from your machine right?

**Comment by NasAmin** [comment]
@julienduchesne Thanks a lot. I'll really appreciate it.

> You are telling me that the current client implementation works from your machine right?

Yes the current implementation works from my machine.

**Comment by NasAmin** [comment]
@julienduchesne Thanks for merging #354 and https://github.com/grafana/grafana-api-golang-client/pull/53. 🎉 
Are you still waiting for other community PRs before creating a release?

**Comment by inkel** [comment]
@NasAmin we are making a final round of pending PRs and opened issues before cutting a new release, but that will be happening soon. Stay tuned!

**Comment by inkel** [comment]
I'm closing this one as #354 was merged.

**Comment by julienduchesne** [comment]
This has been released in v1.19.0
